tests/lax_numpy_ufunc_test.py

```diff
@@ -0,0 +1,287 @@
+  if shape is jtu.PYTHON_SCALAR_SHAPE:
+    return [t for t in dtypes if t in python_scalar_dtypes]
+  return dtypes
+
+def _promote_like_jnp(fun, inexact=False):
+  """Decorator that promotes the arguments of `fun` to `jnp.result_type(*args)`.
+
+  jnp and np have different type promotion semantics; this decorator allows
+  tests make an np reference implementation act more like an jnp
+  implementation.
+  """
+  _promote = _promote_dtypes_inexact if inexact else _promote_dtypes
+  def wrapper(*args, **kw):
+    flat_args, tree = tree_util.tree_flatten(args)
+    args = tree_util.tree_unflatten(tree, _promote(*flat_args))
+    return fun(*args, **kw)
+  return wrapper
+
+UfuncRecord = collections.namedtuple(
+  "UfuncRecord",
+  ["name", "dtypes", "shapes", "rng_factory", "check_dtypes", "inexact", "tol", "reduce_tol"])
+
+def ufunc_record(name, dtypes, shapes, rng_factory, check_dtypes=True, inexact=False, tol=None, reduce_tol=None):
+  return UfuncRecord(name, dtypes, shapes, rng_factory, check_dtypes, inexact, tol=tol, reduce_tol=reduce_tol)
+
+
+UFUNC_RECORDS = [
+    ufunc_record("add", all_dtypes, all_shapes, jtu.rand_default,
+                 reduce_tol={np.float16: 5E-3, jnp.bfloat16: 5E-2}),
+    ufunc_record("divide", all_dtypes, all_shapes, jtu.rand_nonzero, inexact=True),
+    ufunc_record("floor_divide", default_dtypes + unsigned_dtypes, all_shapes, jtu.rand_nonzero),
+    ufunc_record("maximum", all_dtypes, all_shapes, jtu.rand_default),
+    ufunc_record("minimum", all_dtypes, all_shapes, jtu.rand_default),
+    ufunc_record("multiply", all_dtypes, all_shapes, jtu.rand_default),
+    ufunc_record("subtract", number_dtypes, all_shapes, jtu.rand_default),
+    ufunc_record("true_divide", all_dtypes, all_shapes, jtu.rand_nonzero, inexact=True),
+]
+
+
+class JaxNumpyUfuncTests(jtu.JaxTestCase):
+  def _GetArgsMaker(self, rng, shapes, dtypes, np_arrays=True):
+    def f():
+      out = [rng(shape, dtype or jnp.float_)
+             for shape, dtype in zip(shapes, dtypes)]
+      if np_arrays:
+        return out
+      return [jnp.asarray(a) if isinstance(a, (np.ndarray, np.generic)) else a
+              for a in out]
+    return f
+
+  @parameterized.named_parameters(itertools.chain.from_iterable(
+      jtu.cases_from_list(
+        {"testcase_name": jtu.format_test_name_suffix(rec.name, shapes, dtypes),
+         "rec": rec, "shapes": shapes, "dtypes": dtypes}
+        for shapes in filter(
+          _shapes_are_broadcast_compatible,
+          itertools.combinations_with_replacement(rec.shapes, getattr(jnp, rec.name).nargs))
+        for dtypes in itertools.product(
+          *(_valid_dtypes_for_shape(s, rec.dtypes) for s in shapes)))
+      for rec in itertools.chain(UFUNC_RECORDS)))
+  @jax.numpy_rank_promotion('allow')  # This test explicitly exercises implicit rank promotion.
+  def testCall(self, rec, shapes, dtypes):
+    np_ufunc = getattr(np, rec.name)
+    jnp_ufunc = getattr(jnp, rec.name)
+
+    jnp_op = jnp_ufunc
```

Comment:
```suggestion
    jnp_op = getattr(jnp, rec.name)
```

Any reason to bind the intermediate name `jnp_ufunc`? (I find superfluous intermediate names can make it easy to write bugs, so they stand out to me!)

Reply:
Done, thanks!